In [1]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [3]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
df = pd.read_csv('Resources/lending_data.csv')

# Review the DataFrame
print(df.head())  # Print the first 5 rows
print(df.shape)  # Print the number of rows and columns
print(df.info())  # Print a concise summary of the DataFrame
print(df.describe())  # Print summary statistics for numerical columns

   loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0  
(77536, 8)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 77536 entries, 0 to 77535
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   loan_size    

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [4]:
# Separate the data into labels and features
y = df['loan_status']
X = df.drop('loan_status', axis=1)

# Separate the y variable, the labels
y = df['loan_status']

# Separate the X variable, the features
X = df.drop('loan_status', axis=1)


In [5]:
# Review the y variable Series
print(y.head())  # Print the first 5 rows
print(y.value_counts())  # Print the count of each unique value

0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64
loan_status
0    75036
1     2500
Name: count, dtype: int64


In [6]:
# Review the X variable DataFrame
print(X.head())  # Print the first 5 rows
print(X.shape)  # Print the number of rows and columns
print(X.info())  # Print a concise summary of the DataFrame
print(X.describe())  # Print summary statistics for numerical columns

   loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000  
(77536, 7)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 77536 entries, 0 to 77535
Data columns (total 7 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   loan_size         77536 non-null  float64
 1   interest_rate     77536 non-null  float64
 2

### Step 3: Split the data into training and testing datasets by using `train_test_split`.

In [7]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)


---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [8]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
logistic_model = LogisticRegression(random_state=1)

# Fit the model using training data
logistic_model.fit(X_train, y_train)


LogisticRegression(random_state=1)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [9]:
# Make a prediction using the testing data
y_pred = logistic_model.predict(X_test)


In [10]:
# Import the accuracy_score module
from sklearn.metrics import accuracy_score

# Calculate the accuracy score
accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy Score: {accuracy}")

Accuracy Score: 0.9918489475856377


### Step 3: Evaluate the model’s performance by doing the following:

* Generate a confusion matrix.

* Print the classification report.

In [11]:
# Generate a confusion matrix for the model

# Import the required modules
from sklearn.metrics import confusion_matrix, classification_report

# Generate a confusion matrix
confusion_mat = confusion_matrix(y_test, y_pred)
print("Confusion Matrix:")
print(confusion_mat)


Confusion Matrix:
[[18663   102]
 [   56   563]]


In [14]:
# Print the classification report for the model
report = classification_report(y_test, y_pred)
print("Classification Report:")
print(report)

Classification Report:
              precision    recall  f1-score   support

           0       1.00      0.99      1.00     18765
           1       0.85      0.91      0.88       619

    accuracy                           0.99     19384
   macro avg       0.92      0.95      0.94     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** Based on the confusion matrix and classification report, this is how well the logistic regression model predicts the loan status:

For the healthy loan class (0):

The precision of 1.00 means that all loans predicted as healthy were actually healthy.
The recall of 0.99 means that 99% of the actual healthy loans were correctly predicted.
The f1-score of 1.00 suggests excellent performance in predicting healthy loans.
For the high-risk loan class (1):

The precision of 0.85 means that 85% of loans predicted as high-risk were actually high-risk.
The recall of 0.91 means that 91% of actual high-risk loans were correctly identified.
The f1-score of 0.88 suggests good performance, though slightly lower than the healthy class.
Overall, the model performs exceptionally well in predicting healthy loans, with perfect precision and near-perfect recall. For high-risk loans, it has a good balance of precision and recall, accurately identifying most high-risk loans while minimizing false positives.

The overall accuracy of 0.99 indicates that the model correctly predicted the loan status for 99% of the instances.

In summary, the logistic regression model demonstrates excellent performance in predicting both healthy and high-risk loans, with near-perfect results for the healthy class and good results for the high-risk class.

---